### PR TITLE
Sasarkar/transformers 4.40 patches for stopping criteria

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/habana/transformers/generation/stopping_criteria.py [hunk 1]
```diff
@@ -37,26 +44,50 @@ def gaudi_MaxLengthCriteria_call(
...
+    else:
+        # token_idx is present, so we have static shapes, so using single boolean
+        False
 
 def gaudi_StoppingCriteriaList_call(
     self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs
-) -> torch.BoolTensor:
-    is_done = torch.full((input_ids.shape[0],), 0, device=input_ids.device, dtype=torch.int8)
+) -> Union[torch.BoolTensor, bool]:
+    kwargs["out_type_tensor"] = needs_tensor_output(kwargs.get("token_idx",None), kwargs.get("ignore_eos",True), kwargs.get("eos_token_id",None))
```
Comment (regisss) [inline optimum/habana/transformers/generation/stopping_criteria.py:87]:
```suggestion
    kwargs["needs_tensor_output"] = needs_tensor_output(kwargs.get("token_idx",None), kwargs.get("ignore_eos",True), kwargs.get("eos_token_id",None))
```
Let's simply call this kwarg the same

File: optimum/habana/transformers/generation/utils.py [hunk 2]
```diff
@@ -115,6 +115,13 @@ def incrementor(bucket_size, prompt_len):
             "need_expansion": need_expansion,
         }
 
+def get_final_stopping_crit(x):
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:118]:
```suggestion
def get_final_stopping_criteria(x):
```

File: optimum/habana/transformers/generation/utils.py [hunk 3]
```diff
@@ -115,6 +115,13 @@ def incrementor(bucket_size, prompt_len):
...
+    if type(x) == type(True):
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:119]:
```suggestion
    if isinstance(x, bool):
```

File: optimum/habana/transformers/generation/utils.py [hunk 4]
```diff
@@ -115,6 +115,13 @@ def incrementor(bucket_size, prompt_len):
...
+        return x
+    elif torch.is_tensor(x):
+        return all(x)
+    else:
+        assert False
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:124]:
```suggestion
        raise TypeError(f"The stopping criteria should be either a boolean or a torch.tensor but got {type(x)}.")
```

File: optimum/habana/transformers/generation/utils.py [hunk 5]
```diff
@@ -743,6 +750,7 @@ def generate(
         if generation_config.static_shapes:
             # Pad inputs to have static shapes during generation, this gives better performance than dynamic shapes on HPUs
             # In encoder_decoder models, Inputs are already padded
+            assert generation_config.ignore_eos, "For gaudi we pad input_ids with EOS, so this option isnt available for now"
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:753]: Not sure what you mean here. That `ignore_eos == False` is not compatible with static shapes?

Comment (ssarkar2) [reply optimum/habana/transformers/generation/utils.py]: In some cases pad token and eos are same (like in gpt2). `ignore_eos==False` means a stopping criteria is to look for sentences generation eos token and stop. However if we pad with eos tokens, we will make a false stop. I put this assert under another check now `if generation_config.pad_token_id == generation_config.eos_token_id:`. For some models like llama where eos token != pad token, we can have ignore_eos=False

**Comment by ssarkar2** [comment]
@regisss , do you know why style check isnt running for this PR?

**Review by regisss — APPROVED** [review]

**Comment by regisss** [comment]
> @regisss , do you know why style check isnt running for this PR?

Because it is only called for PRs where the target branch is `main`.

**Pull request merged**